src/main/java/org/opentripplanner/updater/vehicle_rental/datasources/GbfsFeedLoader.java

```diff
@@ -0,0 +1,159 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class for managing the state and loading of complete GBFS datasets, and updating them according to individual feed's
+ * TTL rules.
+ */
+public class GbfsFeedLoader {
+    private static final Logger LOG = LoggerFactory.getLogger(GbfsFeedLoader.class);
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    /** One updater per feed type(?)*/
+    private final Map<GBFSFeedName, GBFSFeedUpdater<?>> feedUpdaters = new HashMap<>();
+
+    private final Map<String, String> httpHeaders;
+
+    public GbfsFeedLoader(String url, Map<String, String> httpHeaders, String languageCode) {
+        this.httpHeaders = httpHeaders;
+        URI uri;
+        try {
+            uri = new URI(url);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException("Invalid url " + url);
+        }
+
+        // Fetch autoconfiguration file
+        GBFS data = fetchFeed(uri, httpHeaders, GBFS.class);
+        if (data == null) {
+            throw new RuntimeException("Could not fetch feed " + uri + " autoconfiguration file");
+        }
+
+        // Pick first language if none defined
+        GBFSFeeds feeds = languageCode == null
+                ? data.getFeedsData().values().iterator().next()
+                : data.getFeedsData().get(languageCode);
+        if (feeds == null) {
+            throw new RuntimeException("Language " + languageCode + " does not exist in feed " + uri);
+        }
+
+        // Create updater for each file
+        for (GBFSFeed feed : feeds.getFeeds()) {
+            GBFSFeedName feedName = feed.getName();
+            if (feedUpdaters.containsKey(feedName)) {
+                throw new RuntimeException("Feed " + url + " contains duplicate url for feed " + feedName);
+            }
+            feedUpdaters.put(feedName, new GBFSFeedUpdater<>(feed));
+        }
+    }
+
+    public boolean update() {
+        boolean didUpdate = false;
+
+        for (GBFSFeedUpdater<?> updater : feedUpdaters.values()) {
+            if (updater.shouldUpdate()) {
+                updater.fetchData();
+                didUpdate = true;
+            }
+        }
+
+        return didUpdate;
+    }
+
+    public <T> T getFeed(Class<T> feed) {
+        var updater = feedUpdaters.get(GBFSFeedName.fromClass(feed));
+        if (updater == null) { return null; }
+        return feed.cast(updater.getData());
+    }
+
+    private class GBFSFeedUpdater<T> {
+        private final URI url;
+        private final Class<T> implementingClass;
+
+        private int nextUpdate;
+        private T data;
+
+        private GBFSFeedUpdater(GBFSFeed feed) {
+            url = feed.getUrl();
+            implementingClass = (Class<T>) feed.getName().implementingClass();
```

Comment:
GBFSFeed is a result of directly deserializing the nested object inside the autoconfiguration file, while GBFSFeedName is the enum of the different types, and thus where it is good to place information about the different types. We could add a static method in GBFSFeed to get this, but I'm not sure if that would be any cleaner.

Reply:
What about renaming GBFSFeedName as it seems semantically weird that name can contain something unrelated to name, although I guess name in this case ?

Reply:
I added an additional comment here to clarify the situation. `implemetingClass` is a field on the enum `GBFSFeedName `, and it points to the class, which is used to deserialize the feed with that type. The `getMethod` is a method, which exists in `java.lang.Class`, and is used to find the method reference for that specific implementing class, since they do not share any parent classes, which could be used.

https://github.com/opentripplanner/OpenTripPlanner/pull/3608/commits/a0046f8d1e7f8abfca43a4cfc2ee1fba21bfda0f#diff-cd5ad48285b5873413d48bf43ac6b939274241086a051366d2684aee8e6aa85aR170-R171

Reply:
Ok, this is fine.